### ISSUE TITLE
[Security] Allow to switch user when already switching

### ISSUE THREAD

**Issue by ogizanagi** [symfony/symfony#34690]
**Description**  

The current way this feature works, an authenticated user already impersonating another needs to exit impersonation before switching to another user. Otherwise, he'll hit:

https://github.com/symfony/symfony/blob/2b97a0f5f1ffd2938f2a1a68f333ba9292a270d0/src/Symfony/Component/Security/Http/Firewall/SwitchUserListener.php#L134

First thing: I don't think a `LogicException` is the most suitable here, as it'll generate a 500.
Secondly, is there reasons to forbid this by default?

Then, what about allowing to configure this on the switch user listener? Different strategies might exist:
1. reject, with an appropriate 4XX exception
1. generate a response suggesting to exit & switch to the new user, out-of-the-box.
1. just switch, seamlessly.

Throwing a dedicated exception might allow to perform the second option in userland, but would miss an easy way to switch to the new user once exit.

**Example**  

- Could be a `switch_user` listener option:

```yml
security:
    firewalls:
        # [...]
        main:
            # [...]
            switch_user:
                already_switching_strategy: ~ # one of reject, confirm or switch; reject as default.
```

- a query parameter to force switching
- or the custom exception coupled with error handling in userland. Would miss a native way to exit+switch to the new user though.

**Comment by noniagriconomie** [comment]
@ogizanagi nice idea, i also would like to know the original reason :)

actualy, I do have an exception listener that catch this "feature/problem"
it redirects the user while exiting the switch, and prepopulating for the next switch user (we do have a specific controller where our admin can search a username to launch the switch feature)

It could be very nice to avoid "hack" like that :)

```php

public function onKernelException(GetResponseForExceptionEvent $event)
    {
        $request = $event->getRequest();
        if (!$request->get($this->usernameParameter)) { // query param of the switch feature
            return;
        }

        $exception = $event->getException();
        if (!($exception instanceof \LogicException)) { // ex of switch feature
            return;
        }

        // Switch back programmatically while redirecting on the user switcher controller.
        // We send as well the next username to ease selection for end user.
        $event->setResponse(new RedirectResponse(
            $this->router->generate(
                'front_user_user_switcher_index', [
                    $this->usernameParameter => '_exit', // end switch mode while redirecting
                    '_next_username' => $request->get($this->usernameParameter),
                ]
            ), RedirectResponse::HTTP_FOUND)
        );
        $event->stopPropagation();
    }

```

**Comment by ogizanagi** [comment]
@noniagriconomie: I indeed already use a similar hack for now ;)

**Comment by javiereguiluz** [comment]
For reference purposes, we discussed about this five years ago in #12797.

**Comment by noniagriconomie** [comment]
Thank you for the ref :)

**Comment by wouterj** [comment]
Unfortunately, as with most things in the Security component, these lines of code go all the way back to f216f313e8909015fe9961a5604d179f64c35a90 (introducing the whole component). So there is no reference discussion/PR/issue about this. The only thing I found was #2554, which allowed switching to the user you already are.

Tl;dr: If we do not have a good reason for this exception (and I can't think of any at this moment), we should remove this imho (this applies to many logic in the Security component). I like your suggestion, as this allows this change to be BC:

* If we make a new `AlreadySwitchingException extends LogicException`, we stay BC and strategy (1) & (2) are doable (by using a custom exception listener)
* If we then add a `$allowSeamlessSwitch = false` parameter to the constructor, we can configure whether or not to throw such an exception.

**Comment by ogizanagi** [comment]
Just to complete about:

> If we make a new AlreadySwitchingException extends LogicException, we stay BC

I'm pretty sure the BC promise doesn't apply to logic exceptions anyway. Nobody should rely on such exceptions as it is supposed to identify a wrong code path that needs to be fixed by the developper (which makes this exception irrelevant here, as it can be reached, under legitimate conditions, and there is nothing the developper can do about).
So if an `AuthenticationException` is a better fit, BC is not an issue here :)

**Comment by chalasr** [comment]
https://twitter.com/nikita_ppv/status/1042676176836341760:

> Catching Error exceptions only makes sense in a few contexts, such as top-level exception handlers or unit testing frameworks. 
[...]
LogicException is pretty much the legacy (pre-PHP7) way of signalling this. The intention of Error and LogicException is the same

Changing the exception type is not an issue for me either. 👍 for this

**Comment by clemherreman** [comment]
> I'm pretty sure the BC promise doesn't apply to logic exceptions anyway. Nobody should rely on such exceptions as it is supposed to identify a wrong code path that needs to be fixed by the developper (which makes this exception irrelevant here, as it can be reached, under legitimate conditions, and there is nothing the developper can do about).
> So if an `AuthenticationException` is a better fit, BC is not an issue here :)

I'd vote against. Yes, `\LogicException` is not the best choice, so we might decide that it's not a BC, as there was no conscious choice, and that 

>  Nobody should rely on such exceptions as it is supposed to identify a wrong code path that needs to be fixed by the developper

I agree on that last sentence, however, approaching BC, I usually stick with one rule: *could that change break something that was working before?*

In that case, the answer is clear: *yes, it may*. 

I suggest we treat such a change as a BC, and postpone it to the next major.

**Comment by stephanvierkant** [comment]
How would something break if a `\LogicException` gets replaces by a new `AlreadySwitchingException` that extends `\LogicException`?
